46.md

```diff
@@ -12,51 +12,53 @@ This NIP describes a method for 2-way communication between a remote signer and
 
 ## Terminology
 
-- **Local keypair**: A local public and private key-pair used to encrypt content and communicate with the remote signer. Usually created by the client application.
-- **Remote user pubkey**: The public key that the user wants to sign as. The remote signer has control of the private key that matches this public key.
-- **Remote signer pubkey**: This is the public key of the remote signer itself. This is needed in both `create_account` command because you don't yet have a remote user pubkey.
+- **user**: A person that is trying to use Nostr.
+- **client**: A user-facing application that _user_ is looking at and clicking buttons in. This application will send requests to _remote-signer_.
+- **remote-signer**: A daemon or server running somewhere that will answer requests from _client_, also known as "bunker".
+- **client-keypair/pubkey**: The keys generated by _client_. Used to encrypt content and communicate with _remote-signer_.
+- **remote-signer-keypair/pubkey**: The keys used by _remote-signer_ to encrypt content and communicate with _client_. This keypair MAY be same as _user-keypair_, but not necessarily.
+- **user-keypair/pubkey**: The actual keys representing _user_ (that will be used to sign events in response to `sign_event` requests, for example). The _remote-signer_ generally has control over these keys.
 
 All pubkeys specified in this NIP are in hex format.
 
 ## Initiating a connection
 
-To initiate a connection between a client and a remote signer there are a few different options.
+There are two ways to initiate a connection:
 
-### Direct connection initiated by remote signer
+### Direct connection initiated by _remote-signer_
 
-This is most common in a situation where you have your own nsecbunker or other type of remote signer and want to connect through a client that supports remote signing.
-
-The remote signer would provide a connection token in the form:
+_remote-signer_ provides connection token in the form:
 
 ```
-bunker://<remote-user-pubkey>?relay=<wss://relay-to-connect-on>&relay=<wss://another-relay-to-connect-on>&secret=<optional-secret-value>
+bunker://<remote-signer-pubkey>?relay=<wss://relay-to-connect-on>&relay=<wss://another-relay-to-connect-on>&secret=<optional-secret-value>
```

Comment:
we're missing here the user pubkey, which we need for the `connect` method, unless we send an empty pubkey on the connect and the bunker finds it based on the secret, but we should make either one of these scenarios clear:

`bunker://<bunker-pubkey>?relay=<wss://relay-to-connect-on>&relay=<wss://another-relay-to-connect-on>&secret=<optional-secret-value>&pubkey=<user-pubkey>`

```js
{
     "kind": 24133,
     "pubkey": "<client-pubkey>",
    "content": nip04({
        "id": <random_string>,
        "method": "connect",
        "params": [ "<user-pubkey>", "<secret>" ]
     }),
     "tags": [["p", "<bunker-pubkey>"]],
 }

Reply:
the alternative, if we don't know the user pubkey would be 

```js
{
     "kind": 24133,
     "pubkey": "<client-pubkey>",
    "content": nip04({
        "id": <random_string>,
        "method": "connect",
        "params": [ "", "<secret>" ]
     }),
     "tags": [["p", "<bunker-pubkey>"]],
 }
```

and then the client would need to fetch the pubkey from the bunker -- perhaps the `connect` method should return the user's pubkey always instead of the useless `ack`

Reply:
Good point.

So it turns out that this parameter to `"connect"` is completely useless. Either both client and bunker already know `user-pubkey` OR no one knows `user-pubkey` ahead of time and it should be discovered with `get_public_key`. Is this parameter being used for anything currently? I would be surprised if it was.

Reply:
I had similar thoughts and always thought that the only use of the `connect` (which is quite valid) is to check you can actually do things and that the bunker accepts talking to you.

Once we added the `secret`, the `connect` method is the only way to getting your set of permissions applied to your client pubkey.

Reply:
So we can safely replace it with an empty string for now in the spec? If no one is using it. @mikedilger maybe?

Reply:
> perhaps the connect method should return the user's pubkey always instead of the useless ack

I am currently reusing "ack" differently. Nostr-login creates nostrconnect:// string with a "secret=xxx" param which (secret) is then sent as "connect reply" to client in place of "ack". Secret was needed for the same reason it's in the bunker url. Client signals support for this by providing the secret. Just fyi.

> So we can safely replace it with an empty string for now in the spec? If no one is using it. @mikedilger maybe?

IIRC yes gossip was the one checking the validity of the connect's first parameter.

Reply:
> So we can safely replace it with an empty string for now in the spec? If no one is using it. @mikedilger maybe?

Yeah I check if the connect() pubkey matches.  That's not doing anything though, I think I'm just being an annoying bureaucrat requiring compliance and forms in triplicate ;-).... so I can take it out.